user auth features

doncibo opened the pull request:
Finished user auth features, with demo login button

doncibo committed:
user auth features

joannerd commented on client/src/components/nav_comp/DemoLoginButton.js:
```diff
@@ -0,0 +1,27 @@
+import '../../style/loginmodal.css'
+
+function LoginButton(){
+    const username = "yusuke@example.com"
+    const password = "password"
+    const currentUserId = useSelector(state => state.auth.id);
+    const dispatch = useDispatch();
+    
+    const handleSubmit = (e) => {
+        e.preventDefault();
+        dispatch(login(username, password))
+    }
```
I would suggest moving the `username` and `password` into the `handleSubmit` function since you aren't using the variables anywhere outside of the function.
```suggestion
    const currentUserId = useSelector(state => state.auth.id);
    const dispatch = useDispatch();
    
    const handleSubmit = (e) => {
        e.preventDefault();
        const username = "yusuke@example.com";
        const password = "password";
        dispatch(login(username, password))
    }
```

joannerd commented on client/src/components/nav_comp/DemoLoginButton.js:
```diff
@@ -0,0 +1,27 @@
+    if(currentUserId) return <Redirect to="/" />;
+
+    return(
+        <div className="login" onClick={handleSubmit}>
+            <button id="demologin" >Demo Sign in</button>
```
I would also suggest setting the `onClick` listener for the `<button>`, instead of the `<div>`!

```suggestion
        <div className="login">
            <button id="demologin" onClick={handleSubmit}>Demo Sign in</button>
```

doncibo committed:
Update client/src/components/nav_comp/DemoLoginButton.js

add onclick to button

Co-authored-by: Joanna Chen <x.chenj@gmail.com>

doncibo committed:
Update client/src/components/nav_comp/DemoLoginButton.js

move username and password to handleSubmit

Co-authored-by: Joanna Chen <x.chenj@gmail.com>

joannerd commented on client/src/components/nav_comp/LeftNav.js:
```diff
@@ -0,0 +1,34 @@
+    return(
+        <div className="leftnav">
+            <Link to="/profile" id="logo">
+                <svg version="1.0" xmlns="http://www.w3.org/2000/svg"
+                    viewBox="100 110 150.000000 100.000000"
+                    preserveAspectRatio="xMidYMid meet">
+                    <metadata>
+                        Created by potrace 1.10, written by Peter Selinger 2001-2011
+                    </metadata>
+                    <g transform="translate(0.000000,251.000000) scale(0.050000,-0.050000)"
+                        stroke="none">
+                        <path d="M3260 2741 c0 -11 18 -29 40 -41 54 -29 63 -716 10 -760 -63 -52 -32
+                            -70 120 -70 152 0 183 18 120 70 -25 21 -30 93 -30 422 l0 398 -130 0 c-71 0
+                            -130 -9 -130 -19z"/>
+                        <path d="M2798 2454 c-141 -94 -100 -235 96 -328 110 -52 146 -108 104 -158
+                            -38 -47 -163 -29 -188 27 -36 82 -110 60 -110 -33 l0 -79 132 -14 c368 -39
+                            497 241 177 382 -120 53 -159 118 -98 162 63 46 169 12 169 -54 0 -11 23 -19
+                            50 -19 60 0 76 126 18 148 -81 31 -282 12 -350 -34z"/>
+                    </g>
+                </svg>
```
I recommend following this article in moving this `<svg>` element into an image tag (`<img src={} />`)!
https://blog.logrocket.com/how-to-use-svgs-in-react/

doncibo reviewed (commented):
Made changes: moved onClick and handler from div to button; moved username and password variables to handleSubmit

joannerd commented on client/src/App.js:
```diff
@@ -1,29 +1,39 @@
+
     loadUser();
-  }, []);
+  }, [dispatch]);
```
You can actually remove `dispatch` from the dependency array. Check out this [StackOverflow](https://stackoverflow.com/questions/54930197/react-hooks-dispatch-action-from-useeffect) and note that someone said this:
> React guarantees that dispatch function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

joannerd commented on client/src/components/nav_comp/LoginButton.js:
```diff
@@ -0,0 +1,36 @@
+                    className="login-form-outer"
+                    open={open}
+                    onClose={handleClose}
+                    aria-labelledby="simple-modal-title"                        aria-describedby="simple-modal-description"
```
Just a spacing clean-up suggestion!

```suggestion
                    aria-labelledby="simple-modal-title"
                    aria-describedby="simple-modal-description"
```

doncibo committed:
Update client/src/components/nav_comp/LoginButton.js

spacing issue

Co-authored-by: Joanna Chen <x.chenj@gmail.com>

joannerd commented on client/src/components/nav_comp/LoginModal.js:
```diff
@@ -0,0 +1,58 @@
+                                placeholder="Enter Password"
+                            />
+                        </label>
+                        <button id="signin-button" type="submit" onClick={handleSubmit}>Sign in</button>
```
No need to have both `onSubmit={handleSubmit}` in the `<form>` and `onClick={handleSubmit}` in the `<button>`. You can just choose one to keep and remove the other!

joannerd commented on client/src/components/nav_comp/NavBar.js:
```diff
@@ -0,0 +1,32 @@
+import RightNavLogged from './RightNavLogged'
+import '../../style/navbar.css'
+
+function NavEditor(){
+    const currentUserId = useSelector(state => state.auth.id);
+    let RightNavSection = RightNav
+    if(currentUserId){
+        RightNavSection = RightNavLogged
+    }
+    return RightNavSection
+}
+
+function NavBar(){
+    const RightSection = NavEditor();
+    return (
+        <div className="outer-nav">
+            <div className="navbar">
+                <LeftNav />
+                <MiddleNav />
+                <RightSection />
+            </div>
+        </div>
+    )
+}
```
You can shorten up your code by using a ternary operation to conditionally render the `RightNavLogged` or `RightNav` component based on whether the `currentUserId` is truthy: `{currentUserId ? <RightNavLogged /> : <RightNav />}`.

```suggestion
function NavBar(){
    const currentUserId = useSelector(state => state.auth.id);
    return (
        <div className="outer-nav">
            <div className="navbar">
                <LeftNav />
                <MiddleNav />
                {currentUserId ? <RightNavLogged /> : <RightNav />}
            </div>
        </div>
    )
}
```

joannerd commented on client/src/components/nav_comp/RightNavLogged.js:
```diff
@@ -0,0 +1,104 @@
+    const handleLogOut = (event) => {
+      if (anchorRef.current && anchorRef.current.contains(event.target)) {
+          return
+    }
+    setOpen(false);
+    
+    dispatch(logout())
+    return <Redirect to="/login" />
```
Being super knit-picky and just saw this spacing. 😅

joannerd commented on client/src/components/nav_comp/SignUpModal.js:
```diff
@@ -0,0 +1,80 @@
+                <Container className="signup-form" >
+                    Sign Up!
+                    <form onSubmit={handleSubmit}>
+                        <label>
```
I think these `<label>` tags aren't actually being used, since they're empty label tags.

joannerd commented on client/src/index.js:
```diff
@@ -11,9 +13,16 @@ if (process.env.NODE_ENV !== 'production') {
   getCSRFToken();
 }
 
+const store = configureStore();
+if(process.env.NODE_ENV !== 'production'){
+  window.store = store;
+}
+
```
Nice! I like how you have this for local dev / debugging purposes! 🎉

joannerd commented on client/src/pages/Home.js:
```diff
@@ -0,0 +1,25 @@
+import {LoggedIn, LoggedOut} from './LoggedState'
+import '../style/home.css'
+
+function NavEditor(){
+    const currentUserId = useSelector(state => state.auth.id);
+    let log = LoggedOut
+    if(currentUserId){
+        log = LoggedIn
+    }
+    return log
+}
+
+function Home(){
+    const Log = NavEditor()
+    return (
+        <>  
+            <Log />
+        </>
+    )
+}
```
You can also use the ternary operator to conditionally render `LoggedOut` and `LoggedIn` here.

joannerd commented on client/src/pages/LoginPage.js:
```diff
@@ -0,0 +1,54 @@
+    if(currentUserId) return <Redirect to="/profile" />;
+
+    return(
+        <>
```
No need for the React Fragment since you have a parent `<div>` element wrapping all the inner elements.

joannerd commented on client/src/pages/Pages.js:
```diff
@@ -0,0 +1,20 @@
+        <>  
+            <NavBar />
+            <div className="root-container">
+                <Route path='/login' component={LoginPage} />
+                <Route path='/signup' component={SignUp} />
+                <Route path='/' component={Home} />
```
I heard that some people stick with using single quotes in JavaScript code and double quotes for JSX code, so I started following that and suggest you do as well!

```suggestion
                <Route path="/login" component={LoginPage} />
                <Route path="/signup" component={SignUp} />
                <Route path="/" component={Home} />
```

joannerd commented on client/src/pages/Profile.js:
```diff
@@ -0,0 +1,32 @@
+import { Route } from 'react-router-dom';
+
+
+function NavEditor(){
+    const currentUsername = useSelector(state => state.auth.username);
+    return currentUsername
+    
+}
+function Profile(){
+    let currentUser = NavEditor()
+    return (
+        <>  
+            <Route path="/profile" component={Profile} />
+            <div className="main" >
+                <div className="contentsBanner" >
+                    <div className="left-banner"></div>
+                    <div className="right-banner">
+                        <div id="textwrap" >{currentUser}</div></div>
+                </div>
+                <div className="container-2" >
+                    <div id="leftbuffer" ></div>
+                    <div id="rightbuffer"></div>
+                </div>
+            </div>
+        </>
+    )
+}
```
Just a suggestion for cleaning up this code!
```suggestion
function Profile(){
    const currentUsername = useSelector(state => state.auth.username);
    return (
        <>  
            <Route path="/profile" component={Profile} />
            <div className="main" >
                <div className="contentsBanner" >
                    <div className="left-banner"></div>
                    <div className="right-banner">
                        <div id="textwrap" >{currentUsername}</div></div>
                </div>
                <div className="container-2" >
                    <div id="leftbuffer" ></div>
                    <div id="rightbuffer"></div>
                </div>
            </div>
        </>
    );
}
```

doncibo committed:
Update client/src/components/nav_comp/NavBar.js

well why didn't I think of that? thanks >:D

Co-authored-by: Joanna Chen <x.chenj@gmail.com>

joannerd commented on client/src/pages/SignUp.js:
```diff
@@ -0,0 +1,78 @@
+    const handleSubmit = (e) => {
+        e.preventDefault();
+        dispatch(signup(username, email, password, password2))
+        dispatch(login(username, password))
```
You might need to test this locally to check, but I don't think `dispatch(login(username, password))` is needed.

It looks like when you dispatch the `signup` function, you're return `action.user` in the `authReducer`, meaning that the `state.auth.id` should be updated and cause the `if (currentUserId) return <Redirect to="/profile" />;` line of code to run.

Let me know if it doesn't actually redirect you after removing the `login` dispatch call!

doncibo committed:
Update client/src/pages/Pages.js

bruh lmao, dont you got like 30 other people you gotta do this for, up here changin quotes and stuff jk jk. thanks again ;D

Co-authored-by: Joanna Chen <x.chenj@gmail.com>

doncibo committed:
Update client/src/pages/Profile.js

thanks, yeh that was excessive to have that func on the outside

Co-authored-by: Joanna Chen <x.chenj@gmail.com>
